server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -0,0 +1,87 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.fetch.subphase;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.ReaderUtil;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.index.mapper.DocumentMapper;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.lookup.SourceLookup;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
+ * retrieves the field values from _source and returns them as document fields.
+ */
+public final class FetchFieldsPhase implements FetchSubPhase {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void hitsExecute(SearchContext context, SearchHit[] hits) {
+        FetchFieldsContext fetchFieldsContext = context.fetchFieldsContext();
+        if (fetchFieldsContext == null || fetchFieldsContext.fields().isEmpty()) {
+            return;
+        }
+
+        DocumentMapper documentMapper = context.mapperService().documentMapper();
+        if (documentMapper.sourceMapper().enabled() == false) {
+            throw new IllegalArgumentException("Unable to retrieve the requested [fields] since _source is " +
+                "disabled in the mappings for index [" + context.indexShard().shardId().getIndexName() + "]");
+        }
+
+        Set<String> fields = new HashSet<>();
+        for (String fieldPattern : context.fetchFieldsContext().fields()) {
+            if (documentMapper.objectMappers().containsKey(fieldPattern)) {
+                continue;
+            }
+            Collection<String> concreteFields = context.mapperService().simpleMatchToFullName(fieldPattern);
+            fields.addAll(concreteFields);
+        }
+
+        SourceLookup sourceLookup = context.lookup().source();
+
+        for (SearchHit hit : hits) {
+            int readerIndex = ReaderUtil.subIndex(hit.docId(), context.searcher().getIndexReader().leaves());
+            LeafReaderContext readerContext = context.searcher().getIndexReader().leaves().get(readerIndex);
+            sourceLookup.setSegmentAndDocument(readerContext, hit.docId());
+
+            Map<String, Object> valuesByField = sourceLookup.extractValues(fields);
+
+            for (Map.Entry<String, Object> entry : valuesByField.entrySet()) {
+                String field = entry.getKey();
+                Object value = entry.getValue();
+                List<Object> values = value instanceof List
+                    ? (List<Object>) value
+                    : List.of(value);
```

Comment:
I could see wanting to switch this to something a little less `instanceof` soon. But this'll get the job done for now!